libgnucash/engine/Account.cpp

```diff
@@ -5417,120 +5419,155 @@ gnc_account_imap_delete_account (GncImportMatchMap *imap,
  * probability of a given account with any single token
  */
 struct TokenAccountsInfo
 {
-    std::vector<AccountTokenCount> accounts;
+    std::vector<AccountTokenProbability> accounts;
     int64_t total_count;
 };
 
-/** holds an account guid and its corresponding integer probability
-  the integer probability is some factor of 10
- */
-struct AccountInfo
-{
-    std::string account_guid;
-    int32_t probability;
-};
-
 static void
 build_token_info(char const * suffix, KvpValue * value, TokenAccountsInfo & tokenInfo)
 {
     if (strlen(suffix) == GUID_ENCODING_LENGTH)
     {
         tokenInfo.total_count += value->get<int64_t>();
         /*By convention, the key ends with the account GUID.*/
-        tokenInfo.accounts.emplace_back(AccountTokenCount{std::string{suffix}, value->get<int64_t>()});
+        tokenInfo.accounts.emplace_back(AccountTokenProbability{std::string{suffix}, value->get<int64_t>(), 0.0});
     }
 }
 
-/** We scale the probability values by probability_factor.
-  ie. with probability_factor of 100000, 10% would be
-  0.10 * 100000 = 10000 */
-static constexpr int probability_factor = 100000;
-
-static FinalProbabilityVec
-build_probabilities(ProbabilityVec const & first_pass)
+static TokenInfoVec
+get_token_info(GncImportMatchMap * imap, GList * tokens)
 {
-    FinalProbabilityVec ret;
-    for (auto const & first_pass_prob : first_pass)
+    TokenInfoVec ret;
+    for (auto current_token = tokens; current_token; current_token = current_token->next)
     {
-        auto const & account_probability = first_pass_prob.second;
-        /* P(AB) = A*B / [A*B + (1-A)*(1-B)]
-         * NOTE: so we only keep track of a running product(A*B*C...)
-         * and product difference ((1-A)(1-B)...)
-         */
-        int32_t probability = (account_probability.product /
-                (account_probability.product + account_probability.product_difference)) * probability_factor;
-        ret.push_back({first_pass_prob.first, probability});
+        /* query token frequencies from import match map */
+        TokenAccountsInfo token_info{};
+        std::string token_name{static_cast <char const*> (current_token->data)};
+        auto path = std::string(IMAP_FRAME_BAYES "/") + token_name + "/";
+        qof_instance_foreach_slot_prefix (QOF_INSTANCE (imap->acc), path, &build_token_info, token_info);
+
+        /* calculate conditional account probabilities P(A|T) of account A given token T */
+        if (token_info.total_count > 0)
+        {
+            for (auto & account : token_info.accounts)
+                account.probability = (double)account.token_count / (double)token_info.total_count;
+            ret.emplace_back(token_name, token_info);
+        }
     }
     return ret;
 }
 
-static AccountInfo
-highest_probability(FinalProbabilityVec const & probabilities)
+/** Threshold for conditional account probabilities given individual tokens */
+static constexpr double token_account_probability_threshold = .90; /* 90% */
```

Comment:
Does this need to be `static` if it's already `constexpr`?

Reply:
Either `static` or in the anonymous namespace to keep the symbol from being global.